lib/std/zig/system.zig

```diff
@@ -982,60 +1018,24 @@ fn detectAbiAndDynamicLinker(
-        var buffer: [258]u8 = undefined;
-        while (true) {
-            const file = fs.openFileAbsolute(file_name, .{}) catch |err| switch (err) {
-                error.NoSpaceLeft => unreachable,
-                error.NameTooLong => unreachable,
-                error.PathAlreadyExists => unreachable,
-                error.SharingViolation => unreachable,
-                error.InvalidUtf8 => unreachable,
-                error.BadPathName => unreachable,
-                error.PipeBusy => unreachable,
-                error.FileLocksNotSupported => unreachable,
-                error.WouldBlock => unreachable,
-                error.FileBusy => unreachable, // opened without write permissions
-
-                error.IsDir,
-                error.NotDir,
-                error.InvalidHandle,
-                error.AccessDenied,
-                error.NoDevice,
-                error.FileNotFound,
-                error.NetworkNotFound,
-                error.FileTooBig,
-                error.Unexpected,
-                => |e| {
-                    std.log.warn("Encountered error: {s}, falling back to default ABI and dynamic linker.\n", .{@errorName(e)});
-                    return defaultAbiAndDynamicLinker(cpu, os, query);
-                },
-
-                else => |e| return e,
-            };
-            errdefer file.close();
-
-            const len = preadMin(file, &buffer, 0, buffer.len) catch |err| switch (err) {
-                error.UnexpectedEndOfFile,
-                error.UnableToReadElfFile,
-                => break :blk file,
-
-                else => |e| return e,
-            };
-            const newline = mem.indexOfScalar(u8, buffer[0..len], '\n') orelse break :blk file;
-            const line = buffer[0..newline];
-            if (!mem.startsWith(u8, line, "#!")) break :blk file;
-            var it = mem.tokenizeScalar(u8, line[2..], ' ');
-            file_name = it.next() orelse return defaultAbiAndDynamicLinker(cpu, os, query);
-            file.close();
+        if (followElfShebangs("/usr/bin/env")) |file| {
+            break :blk file;
         }
+        // If /usr/bin/env is missing, iterate PATH to find another "env" binary.
+        const PATH = std.os.getenv("PATH") orelse {
+            std.log.warn("Could not find /usr/bin/env and PATH environment variable is missing, falling back to default ABI and dynamic linker.\n", .{});
+            return defaultAbiAndDynamicLinker(cpu, os, query);
+        };
+        var path_it = mem.tokenizeScalar(u8, PATH, ':');
+        var path_buf: [std.os.PATH_MAX]u8 = undefined;
+        while (path_it.next()) |path| {
+            var path_alloc = std.heap.FixedBufferAllocator.init(&path_buf);
+            const path_env = std.fs.path.join(path_alloc.allocator(), &.{ path, "env" }) catch continue;
+            const env_file = followElfShebangs(path_env) orelse continue;
+            break :blk env_file;
+        }
```

Comment:
It is possible for the directory to be `PATH_MAX` bytes long and have `env` inside of it, in which case this code will not work correctly.

Instead, each directory needs to be opened, and instead of string concatenation + `openFileAbsolute`, it needs to open "env" from the directory handle.

For the `/usr/bin/env` case, note that `std.fs.cwd()` can be used for the directory handle so that you can make `followElfShebangs` work for both situations.

Reply:
OK, fair!

Required changing the implementation slightly.